metaflow/plugins/kubernetes/kubernetes_decorator.py

```diff
@@ -399,14 +414,61 @@ def _save_package_once(cls, flow_datastore, package):
                 [package.blob], len_hint=1
             )[0]
 
+    @classmethod
+    def _parse_decorator_spec(cls, deco_spec: str):
+        if not deco_spec:
+            return cls()
+
+        valid_options = "|".join(cls.defaults.keys())
+        deco_spec_parts = []
+        for part in re.split(f""",(?=[\s\w]+[{valid_options}]=)""", deco_spec):
+            name, val = part.split("=", 1)
+            if name in {"labels", "node_selector"}:
+                try:
+                    json.loads(val.strip().replace('\\"', '"'))
+                except json.JSONDecodeError:
+                    both = name == "node_selector"
+                    val = json.dumps(
+                        cls.parse_kube_keyvalue_list(val.split(","), both),
+                        separators=(",", ":"),
+                    )
+            deco_spec_parts.append("=".join([name, val]))
+        deco_spec_parsed = ",".join(deco_spec_parts)
+        return super()._parse_decorator_spec(deco_spec_parsed)
+
     @staticmethod
-    def parse_node_selector(node_selector: list):
+    def parse_kube_keyvalue_list(items: List[str], requires_both: bool = True):
         try:
-            return {
-                str(k.split("=", 1)[0]): str(k.split("=", 1)[1])
-                for k in node_selector or []
-            }
+            ret = {}
+            for item_str in items:
+                item = item_str.split("=", 1)
+                if requires_both:
+                    item[1]  # raise IndexError
+                ret[str(item[0])] = str(item[1]) if len(item) > 1 else None
+            return ret
         except (AttributeError, IndexError):
-            raise KubernetesException(
-                "Unable to parse node_selector: %s" % node_selector
-            )
+            raise KubernetesException("Unable to parse kubernetes list: %s" % items)
+
+    @staticmethod
+    def clean_kube_labels(
```

Comment:
@dhpollack what's the intention behind this? Ideally if we are supporting labels we would want to pass them as is to kubernetes.

Reply:
Mutating labels may catch the user off guard when the desired label isn't attached as is - we could rely on runtime checks disallowing certain labels instead - I agree with you that that will lead to an error further down during runtime but will be a better no-surprises approach. Also, we do expect that OPA may intercept the request and not allow certain kinds of labels be attached to the pod - so there is that use case. Would it be useful to jump on a quick 15-30 min call to quickly go over all open questions so that we can push this PR out asap? Let me know if any time [here](https://calendar.app.google/7wjKcswSBVdMgMbNA) works.

Reply:
I have changed this to throw an exception now.